File: e2e/selenium/test/resilience.test.js
const assert = require("assert");
const { until, By } = require("selenium-webdriver");
const { VirtualAuthenticatorOptions, Protocol, Transport } = require("selenium-webdriver/lib/virtual_authenticator");
const { buildDriver, testId, loginViaLocal, BASE_URL } = require("./helpers");

// resilience.test.js は e2e/playwright/tests/resilience.spec.ts の4シナリオを Selenium で実装したもの(2回目のe2e監査で追加)
//
// 「壊れやすいのに見落とされがちな、アプリの土台部分の挙動」(ネットワーク遅延・ブラウザ操作・複数タブ・認証手段の後方互換)を対象にする
describe("ネットワーク遅延時の挙動", function () {
  let driver;

  beforeEach(async function () {
    driver = await buildDriver();
  });

  afterEach(async function () {
    await driver.quit();
  });

  it("/api/tasksの応答が遅い間、「読み込み中...」が表示される", async function () {
    // 【なぜCDP接続を使うか】selenium-webdriver には Playwright の page.route()のような高レベルのリクエスト傍受APIが無い
    //
    // Selenium 4はChromeに対してCDPコマンドを直接
    // 実行できる driver.createCDPConnection() を持っており、chromedp/go-rod と同じ
    // CDPのFetchドメインを使えば同じ「特定URLだけ遅延させる」ことができる
    //
    // 【実機検証で判明】
    // このバージョンのselenium-webdriverのCDPConnectionは、
    // 特定コマンドの応答を待つ`send()`と、応答を待たない`execute()`しか公開しておらず、
    // 「Fetch.requestPausedのような非同期イベントを継続的に受け取る」ための`.on()`のような
    // 公開APIが無い(ソース: node_modules/selenium-webdriver/devtools/CDPConnection.js)
    //
    // 内部で保持している`_wsConnection`(生のWebSocket接続)は公開プロパティではないが
    // アクセス自体は可能なため、そこへ直接`on('message', ...)`することでイベント購読を実現する
    const cdpConnection = await driver.createCDPConnection("page");
    await cdpConnection.send("Fetch.enable", {
      patterns: [{ urlPattern: "*/api/tasks*" }],
    });
    cdpConnection._wsConnection.on("message", (data) => {
      let payload;
      try {
        payload = JSON.parse(data.toString());
      } catch (e) {
        return;
      }
      if (payload.method !== "Fetch.requestPaused") return;
      setTimeout(() => {
        cdpConnection.execute("Fetch.continueRequest", { requestId: payload.params.requestId });
      }, 1500);
    });

    await driver.get(`${BASE_URL}/tasks`);
    await loginViaLocal(driver);

    // ログイン成功→リダイレクト後、再度/api/tasksを呼ぶ実装のため、遅延を維持したままリロードし、
    // その瞬間に「読み込み中...」が見えることを確認する(JS版と同じ考え方)
    await driver.navigate().refresh();
    await driver.wait(until.elementLocated(By.xpath("//*[contains(text(), '読み込み中')]")), 5000);
    await driver.wait(until.elementLocated(testId("task-list")), 10000);
  });
});

describe("ブラウザバック・リロード", function () {
  let driver;

  beforeEach(async function () {
    driver = await buildDriver();
  });

  afterEach(async function () {
    await driver.quit();
  });

  it("タスク一覧→ラベル一覧→戻る、で状態が崩れず認証も維持される", async function () {
    await driver.get(`${BASE_URL}/tasks`);
    await loginViaLocal(driver);

    await driver.get(`${BASE_URL}/labels`);
    assert.ok((await driver.getCurrentUrl()).includes("/labels"));

    await driver.navigate().back();
    await driver.wait(until.elementLocated(testId("task-list")), 15000);
  });

  it("タスク一覧をリロードしても再ログインを要求されない(セッションCookieが効いている)", async function () {
    await driver.get(`${BASE_URL}/tasks`);
    await loginViaLocal(driver);

    await driver.navigate().refresh();
    await driver.wait(until.elementLocated(testId("task-list")), 10000);
    const loginInputs = await driver.findElements(testId("login-email-input"));
    assert.strictEqual(loginInputs.length, 0, "リロード後にログイン画面へ飛ばされている");
  });
});

describe("複数タブでのセッション共有", function () {
  let driver;

  beforeEach(async function () {
    driver = await buildDriver();
  });

  afterEach(async function () {
    await driver.quit();
  });

  it("片方のタブでログインすると、同じCookieを共有するもう片方のタブでも認証済み扱いになる", async function () {
    const tab1 = await driver.getWindowHandle();

    await driver.get(`${BASE_URL}/tasks`);
    await loginViaLocal(driver);

    // Selenium 4の switchTo().newWindow("tab") で、同じブラウザセッション(=同じCookie jar)に
    // 新しいタブを開く(Playwright版のcontext.newPage()に相当)
    await driver.switchTo().newWindow("tab");
    await driver.get(`${BASE_URL}/tasks`);
    await driver.wait(until.elementLocated(testId("task-list")), 10000);

    await driver.close();
    await driver.switchTo().window(tab1);
  });

  it("片方のタブでログアウトすると、もう片方のタブは次のアクセスでログイン画面に戻る", async function () {
    const tab1 = await driver.getWindowHandle();
    await driver.get(`${BASE_URL}/tasks`);
    await loginViaLocal(driver);

    await driver.switchTo().newWindow("tab");
    const tab2 = await driver.getWindowHandle();
    await driver.get(`${BASE_URL}/tasks`);
    await driver.wait(until.elementLocated(testId("task-list")), 10000);

    // tab1でログアウト(bff側のRedisセッションが削除される)
    await driver.switchTo().window(tab1);
    await driver.findElement(testId("logout-button")).click();
    await driver.wait(until.elementLocated(testId("login-email-input")), 15000);

    // tab2はリロードして初めてサーバー側のセッション失効に気づく
    await driver.switchTo().window(tab2);
    await driver.navigate().refresh();
    await driver.wait(until.elementLocated(testId("login-email-input")), 15000);
  });
});

describe("パスキー登録後もパスワード認証が引き続き使える(回帰確認)", function () {
  let driver;

  beforeEach(async function () {
    driver = await buildDriver();
  });

  afterEach(async function () {
    await driver.quit();
  });

  it("パスキーを登録したユーザーが、パスワードでもログインできる", async function () {
    const options = new VirtualAuthenticatorOptions();
    options.setProtocol(Protocol.CTAP2);
    options.setTransport(Transport.INTERNAL);
    options.setHasResidentKey(true);
    options.setHasUserVerification(true);
    options.setIsUserVerified(true);
    await driver.addVirtualAuthenticator(options);

    await driver.get(`${BASE_URL}/tasks`);
    await loginViaLocal(driver);

    await driver.get(`${BASE_URL}/account`);
    await driver.wait(until.elementLocated(testId("passkey-register-button")), 10000);
    await driver.findElement(testId("passkey-register-button")).click();
    await driver.wait(until.elementLocated(testId("passkey-register-success")), 10000);

    await driver.findElement(testId("logout-button")).click();
    await driver.wait(until.elementLocated(testId("login-email-input")), 15000);

    // ここが本題: パスキーボタンではなく、通常通りメールアドレス+パスワードでログインし直す
    await loginViaLocal(driver);
    const list = await driver.findElement(testId("task-list"));
    assert.ok(await list.isDisplayed());
  });
});

File: e2e/playwright-go/resilience_test.go
package e2e

// resilience_test.go は e2e/playwright/tests/resilience.spec.ts の4シナリオを
// playwright-goで実装したもの(2回目のe2e監査で追加)
//
// 「壊れやすいのに見落とされがちな、
// アプリの土台部分の挙動」(ネットワーク遅延・ブラウザ操作・複数タブ・認証手段の後方互換)を対象にする
//
// 【なぜJS版とほぼ1対1で移植できたか】playwright-goはPlaywright本体(Node製ドライバ)を
// Goから操作するバインディングであり、chromedp/go-rodのような「生のCDPを自分で組み立てる」レイヤーが無い
// route()・goBack()・context.newPage()等、JS版が使っている高レベルAPIが
// そのままGo版にも存在するため、シナリオの構造をほぼそのまま移植できる
// (go-rodのresilience_test.goで発生した複数タブまわりの不安定さは、playwright-goでは
// 実機検証の範囲では再現しなかった
// 実際のPlaywrightドライバが多重化を面倒見ているためと思われる)
import (
	"testing"
	"time"

	"github.com/mxschmitt/playwright-go"
)

func TestResilience_SlowNetwork_ShowsLoadingState(t *testing.T) {
	page := newPage(t)

	err := page.Route("**/api/tasks*", func(route playwright.Route) {
		time.Sleep(1500 * time.Millisecond)
		_ = route.Continue()
	})
	if err != nil {
		t.Fatalf("Route()の設定に失敗しました: %v", err)
	}

	if _, err := page.Goto(baseURL + "/tasks"); err != nil {
		t.Fatalf("Goto(/tasks)に失敗しました: %v", err)
	}
	loginViaLocal(t, page, false)

	// ログイン成功→リダイレクト後、再度/api/tasksを呼ぶ実装のため、遅延を維持したままリロードし、
	// その瞬間に「読み込み中...」が見えることを確認する(JS版と同じ考え方)
	if _, err := page.Reload(); err != nil {
		t.Fatalf("Reload()に失敗しました: %v", err)
	}
	loadingText := page.GetByText("読み込み中...")
	if err := loadingText.WaitFor(); err != nil {
		t.Fatalf("「読み込み中...」が表示されなかった: %v", err)
	}
	if err := page.GetByTestId("task-list").WaitFor(playwright.LocatorWaitForOptions{
		Timeout: playwright.Float(10000),
	}); err != nil {
		t.Fatalf("task-listが表示されなかった: %v", err)
	}
}

func TestResilience_BrowserBack_PreservesAuthAndState(t *testing.T) {
	page := newPage(t)

	if _, err := page.Goto(baseURL + "/tasks"); err != nil {
		t.Fatalf("Goto(/tasks)に失敗しました: %v", err)
	}
	loginViaLocal(t, page, false)

	if _, err := page.Goto(baseURL + "/labels"); err != nil {
		t.Fatalf("Goto(/labels)に失敗しました: %v", err)
	}

	if _, err := page.GoBack(); err != nil {
		t.Fatalf("GoBack()に失敗しました: %v", err)
	}
	// 戻った後も再ログインを要求されず、一覧がそのまま表示されることを確認する
	if err := page.GetByTestId("task-list").WaitFor(); err != nil {
		t.Fatalf("戻った後にtask-listが表示されなかった(再ログインを要求された可能性): %v", err)
	}
}

func TestResilience_Reload_KeepsSession(t *testing.T) {
	page := newPage(t)

	if _, err := page.Goto(baseURL + "/tasks"); err != nil {
		t.Fatalf("Goto(/tasks)に失敗しました: %v", err)
	}
	loginViaLocal(t, page, false)

	if _, err := page.Reload(); err != nil {
		t.Fatalf("Reload()に失敗しました: %v", err)
	}
	if err := page.GetByTestId("task-list").WaitFor(); err != nil {
		t.Fatalf("リロード後にtask-listが表示されなかった: %v", err)
	}
	count, err := page.GetByTestId("login-email-input").Count()
	if err != nil {
		t.Fatalf("login-email-inputの件数取得に失敗しました: %v", err)
	}
	if count != 0 {
		t.Errorf("リロード後にログイン画面へ飛ばされている(login-email-inputが%d件存在)", count)
	}
}

// TestResilience_MultiTab_SessionSharedAndLogoutPropagates は
// 「片方のタブでログインすると、同じCookieを共有するもう片方のタブでも認証済み扱いになる」
// 「片方のタブでログアウトすると、もう片方のタブは次のアクセスでログイン画面に戻る」を確認する
//
// 【go-rod版との対比】go-rod版resilience_test.goでは、同じブラウザ内の2つ目のタブへの
// 操作がCDPレベルでハングする問題に何度も当たり、後半のログアウト伝播シナリオを断念した
//
// playwright-go では同じ流れ(page.Context().NewPage()で同一 BrowserContext 内に第2タブを開く)を素直に実装した範囲で、実機検証上は問題無く動いた
func TestResilience_MultiTab_SessionSharedAndLogoutPropagates(t *testing.T) {
	tab1 := newPage(t)
	if _, err := tab1.Goto(baseURL + "/tasks"); err != nil {
		t.Fatalf("tab1 Goto(/tasks)に失敗しました: %v", err)
	}
	loginViaLocal(t, tab1, false)

	// 同じブラウザコンテキスト(=同じCookie jar)から新しいタブを開く
	tab2, err := tab1.Context().NewPage()
	if err != nil {
		t.Fatalf("tab2の作成に失敗しました: %v", err)
	}
	if _, err := tab2.Goto(baseURL + "/tasks"); err != nil {
		t.Fatalf("tab2 Goto(/tasks)に失敗しました: %v", err)
	}
	if err := tab2.GetByTestId("task-list").WaitFor(); err != nil {
		t.Fatalf("tab2(同じブラウザの新タブ)がログイン済み扱いにならなかった: %v", err)
	}

	// tab1でログアウト(bff側のRedisセッションが削除される)
	if err := tab1.GetByTestId("logout-button").Click(); err != nil {
		t.Fatalf("tab1でのログアウトのクリックに失敗しました: %v", err)
	}
	if err := tab1.GetByTestId("login-email-input").WaitFor(); err != nil {
		t.Fatalf("tab1でのログアウト後にログイン画面が表示されなかった: %v", err)
	}

	// tab2はリロードして初めてサーバー側のセッション失効に気づく
	if _, err := tab2.Reload(); err != nil {
		t.Fatalf("tab2のReload()に失敗しました: %v", err)
	}
	if err := tab2.GetByTestId("login-email-input").WaitFor(playwright.LocatorWaitForOptions{
		Timeout: playwright.Float(15000),
	}); err != nil {
		t.Fatalf("tab1でのログアウト後、tab2がログイン画面に戻らなかった: %v", err)
	}
}

// TestResilience_PasskeyRegistration_DoesNotBreakPasswordLogin は
// パスキー機能追加(auth_mode="passkey"分岐等の共通コード変更)が、既存のパスワードログイン経路を
// 壊していないことを、「パスキー登録済み」の状態から改めて確認する回帰テスト
func TestResilience_PasskeyRegistration_DoesNotBreakPasswordLogin(t *testing.T) {
	page := newPage(t)

	// passkey_test.goと同じ生CDPコマンドで仮想認証器を有効化(playwright-goには
	// WebAuthn専用の高レベルAPIが無いため)
	session, err := page.Context().NewCDPSession(page)
	if err != nil {
		t.Fatalf("CDPSessionの作成に失敗しました: %v", err)
	}
	if _, err := session.Send("WebAuthn.enable", map[string]any{}); err != nil {
		t.Fatalf("WebAuthn.enableに失敗しました: %v", err)
	}
	_, err = session.Send("WebAuthn.addVirtualAuthenticator", map[string]any{
		"options": map[string]any{
			"protocol":                    "ctap2",
			"transport":                   "internal",
			"hasResidentKey":              true,
			"hasUserVerification":         true,
			"isUserVerified":              true,
			"automaticPresenceSimulation": true,
		},
	})
	if err != nil {
		t.Fatalf("WebAuthn.addVirtualAuthenticatorに失敗しました: %v", err)
	}

	if _, err := page.Goto(baseURL + "/tasks"); err != nil {
		t.Fatalf("Goto(/tasks)に失敗しました: %v", err)
	}
	loginViaLocal(t, page, false)

	if _, err := page.Goto(baseURL + "/account"); err != nil {
		t.Fatalf("Goto(/account)に失敗しました: %v", err)
	}
	if err := page.GetByTestId("passkey-register-button").Click(); err != nil {
		t.Fatalf("passkey-register-buttonのクリックに失敗しました: %v", err)
	}
	if err := page.GetByTestId("passkey-register-success").WaitFor(); err != nil {
		t.Fatalf("パスキー登録に失敗しました: %v", err)
	}

	if err := page.GetByTestId("logout-button").Click(); err != nil {
		t.Fatalf("logout-buttonのクリックに失敗しました: %v", err)
	}
	if err := page.GetByTestId("login-email-input").WaitFor(); err != nil {
		t.Fatalf("ログアウト後にログイン画面が表示されなかった: %v", err)
	}

	// ここが本題: パスキーボタンではなく、通常通りメールアドレス+パスワードでログインし直す
	loginViaLocal(t, page, false)
	if err := page.GetByTestId("task-list").WaitFor(); err != nil {
		t.Fatalf("パスキー登録後のパスワードログインでtask-listが表示されなかった: %v", err)
	}
}
